[Fixed] Joker shop price not properly erased on purchase

blackcan1122 opened the pull request:
regarding #277 

Implemented a way to calculate Sprite Size

MeirGavish commented:
@blackcan1122 Thanks for fixing this! Can you add a screenshot/video showing it's fixed? Can you also check the prices under jokers when highlighting them for sale get displayed and erased properly and maybe add screenshots/video for that as well?

blackcan1122 commented:
Video Link for buying 2 digit jokers and 2x 1 digit jokers

https://www.youtube.com/watch?v=9JtdoLBCOPw

ricfehr3 commented on source/sprite.c:
```diff
@@ -278,6 +278,36 @@ void sprite_object_set_focus(SpriteObject* sprite_object, bool focus)
+    OBJ_ATTR* obj = sprite_object->sprite->obj;
+
+    int shape = (obj->attr0 >> 14) & 0x3;
+    int size = (obj->attr1 >> 14) & 0x3;
```
These should use the built in libtonc functions from `tonc_oam.h`:

```c
INLINE const u8 *obj_get_size(const OBJ_ATTR *obj);
INLINE int obj_get_width(const OBJ_ATTR *obj);
INLINE int obj_get_height(const OBJ_ATTR *obj);
```

ricfehr3 commented on source/game.c:
```diff
@@ -3517,22 +3517,29 @@ static void game_round_end_dismiss_round_end_panel()
-    price_rect.right = price_rect.left + TILE_SIZE * 3;
-    // Taking 2 tiles down so the highlighted case is also covered
-    price_rect.bottom = price_rect.top + 2 * TILE_SIZE;
+    // todo: fix the + 4 offset, don't know why its needed
```
Please don't add nebulous magic fixes. The value 4 needs some context to it's purpose in either a variable/macro name or a comment.

ricfehr3 reviewed (changes requested):
Hey, thank you for this PR. I left a couple comments below. Once they are good, I think I'm good to approve. Thanks!

blackcan1122 commented:
https://github.com/user-attachments/assets/4674e1fa-e397-4cb3-afab-821353895940

I figured out where the mysterious “+4” magic number in the erase logic came from.

As a small test, I stored the card’s Y value globally when the shop was generated and then used that Y value to define the erase rectangle. This worked, so it seems that the floating/hover animation also changes the sprite’s Y position when it runs.

To make the erase logic robust, I added the full maximum selection height (10 px) to the bottom of the rectangle. That way, no matter when this function is called (during the animation or after it finishes), we always erase the text correctly.

Edit:

i moved the define macros from sprite.c to sprite.h so they are accesseable in other files, i hope you guys are fine with it

blackcan1122 commented on source/sprite.c:
```diff
@@ -278,6 +275,44 @@ void sprite_object_set_focus(SpriteObject* sprite_object, bool focus)
+    *width = size[0];
+    *height = size[1];
+    return true;
+}
```
Sure no problem! :)

blackcan1122 commented on source/game.c:
```diff
@@ -3517,22 +3517,38 @@ static void game_round_end_dismiss_round_end_panel()
+    {
+        // fallback
+        height = CARD_SPRITE_SIZE;
+        width = CARD_SPRITE_SIZE / 2;
```
I haven't actually tested the fallbacks specifically yet, but due to the name I expected that CARD_SPRITE_SIZE would be square dimensions, so I roughly approximated that half the width of the square would be the width of the card. If you want i can test the Fallbacks too :)

I also don't 100% know when this fallback would trigger in a real scenario, but I like to at least have some visual feedback that something may be broken or unhandled. Since there is no log or anything, I can't print an error to it.
